packages/java-parser/src/comments.js

```diff
@@ -0,0 +1,156 @@
+  }
+
+  // edge case: when the file start with comments, it attaches as leadingComments to the first token
+  const firstToken = tokens[0];
+  const headComments = [];
+  while (
+    attachComments.length > 0 &&
+    attachComments[0].endOffset < firstToken.startOffset
+  ) {
+    headComments.push(attachComments[0]);
+    attachComments.splice(0, 1);
+  }
+
+  if (headComments.length > 0) {
+    firstToken.leadingComments = headComments;
+  }
+
+  // edge case: when the file end with comments, it attaches as trailingComments to the last token
+  const lastToken = tokens[tokens.length - 1];
+  const tailComments = [];
+  while (
+    attachComments.length > 0 &&
+    attachComments[attachComments.length - 1].startOffset > lastToken.endOffset
+  ) {
+    tailComments.push(attachComments[attachComments.length - 1]);
+    attachComments.splice(attachComments.length - 1, 1);
+  }
+
+  if (tailComments.length > 0) {
+    lastToken.trailingComments = tailComments.reverse();
+  }
+
+  let currentToken = 0;
+  attachComments.forEach(element => {
+    // find the correct position to place the comment
+    while (
+      !(
+        element.startOffset > tokens[currentToken].endOffset &&
+        element.endOffset < tokens[currentToken + 1].startOffset
+      )
+    ) {
+      currentToken++;
+    }
+
+    // attach comment to the closest token
+    if (
+      element.startOffset - tokens[currentToken].endOffset <
+      tokens[currentToken + 1].startOffset - element.endOffset
+    ) {
+      if (!tokens[currentToken].hasOwnProperty("trailingComments")) {
+        tokens[currentToken].trailingComments = [];
+      }
+      tokens[currentToken].trailingComments.push(element);
+    } else {
+      if (!tokens[currentToken + 1].hasOwnProperty("leadingComments")) {
+        tokens[currentToken + 1].leadingComments = [];
+      }
+      tokens[currentToken + 1].leadingComments.push(element);
+    }
+  });
+
+  return tokens;
+}
+
+function findUpperBoundToken(tokens, comment) {
+  let diff;
+  let i;
+  let current;
+
+  let len = tokens.length;
+  i = 0;
+
+  while (len) {
+    diff = len >>> 1;
+    current = i + diff;
+    if (tokens[current].startOffset > comment.startOffset) {
+      len = diff;
+    } else {
+      i = current + 1;
+      len -= diff + 1;
+    }
+  }
+  return i;
+}
+
+function extendCommentRange(tokens, comments) {
+  const ignoreComments = [...comments];
+  let position;
+  ignoreComments.forEach(comment => {
+    position = findUpperBoundToken(tokens, comment);
+    comment.extendedRange = {};
+    comment.extendedRange.startOffset =
+      position - 1 < 0 ? comment.startOffset : tokens[position - 1].endOffset;
+    comment.extendedRange.endOffset =
+      position == tokens.length
+        ? comment.endOffset
+        : tokens[position].startOffset;
+  });
+  return ignoreComments;
+}
+
+function filterPrettierIgnore(comments) {
+  return [...comments].filter(comment =>
+    comment.image.match(/\/\/(\s*)prettier-ignore(\s*)/gm)
```

Comment:
can't a prettier-ignore comment be a multi line comment too? (/* prettier ignore */)

Reply:
yes you are right, I forgot that case :+1: